### DISCUSSION TITLE
New way of combining server-side code with client-side code in the `app` dir

### DISCUSSION THREAD

**Discussion by DanieliusDev** [vercel/next.js#48109] [Ideas]
### Goals

1. Allow to use server components that return a client component which handles client-side code.

### Non-Goals

1. Removal of the `"use client"` directive, unless this becomes a out of the box react feature.

By this i mean this being supported:

```js
function MyComponent() {
    // Server-side code here

    return () => {
        // Client-side code here

        return <h1>Some JSX</h1>;
    }
}

// this is currently not allowed because this component doesn't directly return jsx 
<MyComponent />
```

### Background

Currently with the `app` directory all components are server components by default, which is fine. But if we wanted to use client-side code, we are required to add the `"use client"` directive, which is also fine. But.... if we want to use server-side code **and** use client-side code in the same component, we are forced to split our code into two components which one handles the server-side code and the other handles client-side code. We also have to share our fetched data (for example) via a prop to our client-side component, which is not quite ideal. Here is an example:

`app/page.js`:
```js
import Counter from './Counter.js';

export default async function App() {
    const data = await getData();

    return <Counter initialCount={data.count} />;
}
```

`app/Counter.js`
```js
'use client';

export default function Counter({ initialCount }) {
    const [count, setCount] = useState(initialCount);

    return (
        <div>
            <h1>Count: {count}</h1>
            <button onClick={() => setCount(count + 1)}>Click me</button>
        </div>
    );
}
```

This could be majorly simplified for a better developer experience and code readability.

### Proposal

Allow to use server components that return a client component which handles client-side code. Here is an example of how we can simplify the above code:

`app/page.js`
```js
export default async function App() {
    const data = await getData();

    return () => {
        const [count, setCount] = useState(data.count);

        return (
            <div>
                <h1>Count: {count}</h1>
                <button onClick={() => setCount(count + 1)}>Click me</button>
            </div>
        );
    }
}
```

No need for creating a separate component just to handle client-side code, no need for any `"use client"` directive, and no need for passing props. Note that the returned client component should **never** take props and instead use props from the server component if needed. Also a last note, we should still be able to directly return jsx if our component doesn't need client-side code. 

I am happy to help implement this or share further ideas for this if i can think of any.

An implementation would go something like this:

```js
const result = await Component(props);

if (typeof result === 'function') {
    // render result() on the client
} else {
    // render result on the server
}
```

**Comment by suiramdev** [comment]
I just encountered the same problem while coding my CMS so I decided to go back to the `page/` directory. I also don't think you should split the page into two components just to use the states. It makes the folder structure more blurry and can potentially introduce other problems.

In my experience, it is simpler and more straightforward to use the states directly in the page component. This also keeps the logic in one place, which can make it easier to manage.

In my opinion, sometimes it is better to prioritize the client side rather than the server side... It depends on the project.
The directive that Next.JS takes is interesting for sites, not for some applications (CMS, ...)

**Reply by suiramdev** [reply]
This is especially the case when our page is intended to be only client-side.

![](https://pbs.twimg.com/media/FtOjOmxXsAYimZk?format=jpg&name=4096x4096)

**Reply by DanieliusDev** [reply]
This feature request aims to balance client-side and server-side and not to prioritise one of them.

**Reply by DanieliusDev** [reply]
Imo, every react component should be a server component which you can enable client functionality via returning a function (client component **without props**).

**Reply by suiramdev** [reply]
> Imo, every react component should be a server component which you can enable client functionality via returning a function (client component **without props**).

I find your idea interesting too. I just wanted to support that this problem is common for this kind of applications and that it should be fixed somehow.
